### PR TITLE
Add support for In-Reply-To parameter in mailto: URIs

### PR DESCRIPTION
**Pull request by mehdisadeghi**
According to the RFC In-Reply-To header could contain
multiple addresses with flexible format. K-9 creently
supports only a single In-Reply-To, so I assumed a
single In-Reply-To for mailto: links, to keep it simple.

Fixes https://github.com/k9mail/k-9/issues/5193
Initial disucssion at: https://forum.k9mail.app/t/in-reply-to-header-not-set/766

### CODE REVIEW

File: mail/common/src/main/java/com/fsck/k9/helper/MailTo.java
Comment (cketti) [inline mail/common/src/main/java/com/fsck/k9/helper/MailTo.java:70]: We'll have to perform some input validation here. Skipping this step could lead to being able to inject arbitrary headers or to an invalid message being created. We have code in place to avoid both of those scenarios. But triggering this code means message creation will fail (currently without letting the user know what went wrong exactly). We still want users to be able to compose valid messages from mailto: URIs that contain (some) invalid fields.

To accomplish this you can use `MessageIdParser.parseList()` once PR #5195 is merged. Please only use the first item of the returned list and if the method throws any `MimeHeaderParserException`s write `null` to `inReplyToMessageId`.

**Review by cketti — CHANGES_REQUESTED** [review]
Thanks for the pull request :heart:

It requires just a few small changes.

**Comment by cketti** [comment]
Please also add at least two test cases to `MailToTest`. One with a valid `in-reply-to` parameter and one with an invalid one.

File: mail/common/src/main/java/com/fsck/k9/helper/MailTo.java
Comment (mehdisadeghi) [reply mail/common/src/main/java/com/fsck/k9/helper/MailTo.java:70]: Sure, makes sense. And thanks for the other PR. I'll update as soon as the other one is merged.

Comment (cketti) [reply mail/common/src/main/java/com/fsck/k9/helper/MailTo.java:70]: I merged #5195.

Comment (mehdisadeghi) [reply mail/common/src/main/java/com/fsck/k9/helper/MailTo.java:70]: @cketti I just tried to apply your comments and found a few interesting things. First of all your parser which you added so quickly looks great! Second part goes to the problems of course :)
1. I just learned that mailto: has its own rfc: [RFC 6808](https://tools.ietf.org/html/rfc6068).
2. Header field values in a mailto: links according to this RFC are case-incensitive, so no difference between in-reply-to and In-Reply-To ([page 5](https://tools.ietf.org/html/rfc6068#page-5)) in a mailto: link.
3. The `<` and `>` are not present in `mailto:` links ([page 3](https://tools.ietf.org/html/rfc6068#page-3)).

So in short it looks like this:

      mailtoURI    = "mailto:" [ to ] [ hfields ]
      to           = addr-spec *("," addr-spec )
      hfields      = "?" hfield *( "&" hfield )
      hfield       = hfname "=" hfvalue
      hfname       = *qchar
      hfvalue      = *qchar

Currently the parser expects  `<` and `>` in mailto: links which is rejecting valid mailto: links. So I can't use it for parsing mailto: links. Is it possible to adapte the parse to also understand mailto: links or a MailToParser is necessary?
Thanks.

File: app/ui/legacy/src/main/java/com/fsck/k9/activity/MessageCompose.java
Comment (mehdisadeghi) [reply app/ui/legacy/src/main/java/com/fsck/k9/activity/MessageCompose.java:1487]: Move it to the MailTo class.

File: mail/common/src/main/java/com/fsck/k9/helper/MailTo.java
Comment (cketti) [reply mail/common/src/main/java/com/fsck/k9/helper/MailTo.java:70]: > Header field values in a mailto: links according to this RFC are case-incensitive, so no difference between in-reply-to and In-Reply-To ([page 5](https://tools.ietf.org/html/rfc6068#page-5)) in a mailto: link.

Our `MailTo` class already takes that into account via `CaseInsensitiveParamWrapper`.


> The `<` and `>` are not present in `mailto:` links ([page 3](https://tools.ietf.org/html/rfc6068#page-3)).

That page only talks about `addr-spec`, which is not relevant for this case. Angle brackets are not valid characters in URIs and so have to be encoded as %3C and %3E respectively. You can also see this in an example in section 6.1:

`mailto:list@example.org?In-Reply-To=%3C3469A91.D10AF4C@example.com%3E`

Our `MailTo` class already handles decoding percent-encoded values (the `Uri` class is doing the heavy lifting for us). Handing above URI to `MailTo.parse()` and then calling `params.getQueryParameters("in-reply-to")` will return a list containing `<3469A91.D10AF4C@example.com>` as the sole element. You can pass that string to `MessageIdParser`  as-is to make sure it is valid.

Comment (mehdisadeghi) [reply mail/common/src/main/java/com/fsck/k9/helper/MailTo.java:70]: > That page only talks about addr-spec, which is not relevant for this case. 

Yes, however it also defines `mailtoURI    = "mailto:" [ to ] [ hfields ]` and `hfield(s)` is defined further as `hfield       = hfname "=" hfvalue` which in turn defines `hfname` and `hfvalue` as `*qchar`. So it will apparently accept triangular brackets but is also valid without them. And that's where my problem is, mailto: links with in-reply-to but without triangular brackets.

> Angle brackets are not valid characters in URIs and so have to be encoded as %3C and %3E respectively. You can also see this in an example in section 6.1

Right, I had overlooked that.

So I'm not sure whether I have to manually add triangular brackets to strings which don't contain them before sending them to parser? I'm looking into it now.

Comment (mehdisadeghi) [reply mail/common/src/main/java/com/fsck/k9/helper/MailTo.java:70]: As an example see "Reply to thread" buttons on sourcehut: https://lists.sr.ht/~sircmpwn/sr.ht-discuss/%3CC9Q4WQUDIYWZ.3IC0HPAWPPS8H%40debian-alex%3E

Comment (cketti) [reply mail/common/src/main/java/com/fsck/k9/helper/MailTo.java:70]: > > That page only talks about addr-spec, which is not relevant for this case.
> 
> Yes, however it also defines `mailtoURI = "mailto:" [ to ] [ hfields ]` and `hfield(s)` is defined further as `hfield = hfname "=" hfvalue` which in turn defines `hfname` and `hfvalue` as `*qchar`. So it will apparently accept angular brackets but is also valid without them. And that's where my problem is, mailto: links with in-reply-to but without angular brackets.

`hfvalue` may not contain angle brackets. However it refers to the header field value in its percent-encoded form. In the example `%3C3469A91.D10AF4C@example.com%3E` does not contain angle brackets and hence matches `hfvalue`. After percent-decoding it does (and it must, because RFC 5322 defines `In-Reply-To` that way). The mailto RFC does not specify special handling for `In-Reply-To`, so RFC 5322 applies (plus RFC 6068 encoding rules).
 
> So I'm not sure whether I have to manually add triangular brackets to strings which don't contain them before sending them to parser? I'm looking into it now.

No, you shouldn't modify the value in any way. Sourcehut is using an `in-reply-to` value that is not valid according to RFC 5322. With the check I asked you to add K-9 Mail would simply ignore it.

Comment (cketti) [reply mail/common/src/main/java/com/fsck/k9/helper/MailTo.java:70]: > Sourcehut is using an `in-reply-to` value that is not valid according to RFC 5322. With the check I asked you to add K-9 Mail would simply ignore it.

That's not entirely correct. It most likely is valid when taking the obsolete syntax (see `obs-in-reply-to`) into account. But it's 2021, nobody should be generating links like that anymore.

Comment (mehdisadeghi) [reply mail/common/src/main/java/com/fsck/k9/helper/MailTo.java:70]: Thanks for the update. I'll discuss this right a way on Sourcehut and come back with more information. However, IMHO it is worth it to remain backward compatible if it does not have any substantial overhead and is easy to implement.

Comment (cketti) [reply mail/common/src/main/java/com/fsck/k9/helper/MailTo.java:70]: I'm happy to support the obsolete syntax to be able to properly display old emails, say from an archive. However, we try hard not to generate emails that use the obsolete syntax. 
Adding support for reading mailto: URIs using the obsolete syntax doesn't do any good because the read values are used to generate a new email.

Also, the RFC contains no clear information on whether `abc@domain.example` and `<abc@domain.example>` are the same message identifier. Because of this I'd argue they are two distinct message identifiers and adding angle brackets around the first version would be wrong.

All of this ambiguity goes away if we don't allow the obsolete syntax in mailto: URIs.

Comment (mehdisadeghi) [reply mail/common/src/main/java/com/fsck/k9/helper/MailTo.java:70]: > I'm happy to support the obsolete syntax to be able to properly display old emails, say from an archive.

Glad to hear that!

> Adding support for reading mailto: URIs using the obsolete syntax doesn't do any good because the read values are used to generate a new email.

I'm still waiting for a reply from sr.ht, however let's look at the problem from a different perspective:

Does implementation details of this patch affect the current functionality of K-9 when _not opened_ via mailto:?
No.
Does implementation details of this patch affect the current functionality of K-9 when _opened_ via mailto:?
No. (except for error handling, and consider that the feature does not exists at the moment)
Does the format of in-reply-to within mailto: affect how In-Reply-To is handled when replying to mailing lists?
No.

Considering the above questions, isn't it much better to simply take the in-reply-to value from mailto: and wrap it in angle brackets if necessary and use it?
I mean, the way in-reply-to is represented in mailto: is isolated from email headers because in-reply-to message-id should anyway be wrapped in angle brackets to be added to mail headers (RFC 5322). So why to be strict here when it is possible to support old and new syntax at the same time?

**Comment by mehdisadeghi** [comment]
> Please also add at least two test cases to `MailToTest`. One with a valid `in-reply-to` parameter and one with an invalid one.

I added a few tests.

File: mail/common/src/main/java/com/fsck/k9/helper/MailTo.java
Comment (mehdisadeghi) [reply mail/common/src/main/java/com/fsck/k9/helper/MailTo.java]: Sourcehut now uses angle brackets in mailto: links! However, my previous arguement still holds, but at this point is only a nice to have for me :)

File: mail/common/src/main/java/com/fsck/k9/helper/MailTo.java
Comment (cketti) [inline mail/common/src/main/java/com/fsck/k9/helper/MailTo.java:84]: When `MessageIdParser.parseList()` doesn't throw an exception it always returns a list with at least one valid message identifier. So this code can be simplified to:

```suggestion
                inReplyToMessageId = inReplyToMessageIds.get(0);
```

File: mail/common/src/test/java/com/fsck/k9/helper/MailToTest.java
Comment (cketti) [inline mail/common/src/test/java/com/fsck/k9/helper/MailToTest.java:198]: You mixed up the order of arguments here and in the other tests. It's `assertEquals(expected, actual)`.

File: mail/common/src/test/java/com/fsck/k9/helper/MailToTest.java
Comment (cketti) [inline mail/common/src/test/java/com/fsck/k9/helper/MailToTest.java:203]: Message identifiers are not separated by semicolons, which is why `MessageIdParser` doesn't like this. If you remove the semicolon or replace it with a space (`%20` when encoded) your code will properly extract the first message identifier.

File: mail/common/src/test/java/com/fsck/k9/helper/MailToTest.java
Comment (cketti) [inline mail/common/src/test/java/com/fsck/k9/helper/MailToTest.java:194]: By not encoding `<` as `%3C` and `>` as `%3E` you're relying on `Uri.parse()` accepting a URI with an invalid syntax. I don't mind having one such test, but it should be spelled out in the method name what is being tested.
For all other tests please use `%3C` and `%3E`.

File: mail/common/src/test/java/com/fsck/k9/helper/MailToTest.java
Comment (cketti) [inline mail/common/src/test/java/com/fsck/k9/helper/MailToTest.java:10]: The import is unused.

**Review by cketti — CHANGES_REQUESTED** [review]
Almost there :+1:

**Review by KunKaxx — COMMENTED** [review]
.

File: mail/common/src/main/java/com/fsck/k9/helper/MailTo.java
Comment (cketti) [reply mail/common/src/main/java/com/fsck/k9/helper/MailTo.java]: I was wrong. Even the obsolete syntax required angle brackets.

Supporting invalid forms only leads to the proliferation of these invalid forms. If enough implementations add support for it because it occurs in the wild we've implicitly created a new standard. One that isn't documented anywhere except via the various implementations. That's not a desirable situation. 

Being strict leads to buggy implementations being discovered and hopefully fixed. So that's what we aim for.

**Review by KunKaxx — COMMENTED** [review]
.

File: mail/common/src/main/java/com/fsck/k9/helper/MailTo.java
Comment (mehdisadeghi) [reply mail/common/src/main/java/com/fsck/k9/helper/MailTo.java]: Wow, thanks for the great explanation! I'm convinced! 😄

**Comment by mehdisadeghi** [comment]
> Almost there +1

Thanks for the thorough review! All applied.

**Review by cketti — APPROVED** [review]

**Comment by cketti** [comment]
Thanks :+1:

**Pull request merged**
